### DISCUSSION TITLE
Close/Recovery alerts issue

### DISCUSSION THREAD

**Discussion by RanjithPremanand** [grafana/grafana#37025] [Alerting]
**What happened**:

We have integrated grafana with opsgenie as an alert notification channel. Whenever there is an alert triggered from grafana to opegenie, it creates an alert in opsgenie based on the create alert action setup as expected but when the alert goes back to normal state, ideally the opsgenie alert should also get closed based on the close alert action setup, but it is not doing so. Tried reaching opsgenie support and they said in grafana the script should be edited in such a way to have the Close/Recovery alerts send the same fields over as when it sends the Create Alert actions.

**What you expected to happen**:

Close/Recovery alerts should send the same fields similar to when it sends the Create Alert actions.

**How to reproduce it (as minimally and precisely as possible)**:

Integrating grafana with opsgenie as a notification channel and by triggering an alert in grafana will create an alert in opsgenie. Similarly once the alert goes back to normal in grafana, alert will not get auto closed in opsgenie due to this issue.

**Environment**:
- Grafana version: v7.0.3
- Data source type & version: Prometheus, version 2.18.2
- OS Grafana is installed on: Running as a pod in GCP kubernetes cluster
- User OS & Browser: MAC & Chrome

<img width="540" alt="Ref-3" src="https://user-images.githubusercontent.com/80676553/111132068-eb10b600-859e-11eb-9f6b-6b2a093b8962.png">
<img width="546" alt="Ref-1" src="https://user-images.githubusercontent.com/80676553/111132094-f2d05a80-859e-11eb-881f-c27afb46700f.png">
<img width="532" alt="Ref-2" src="https://user-images.githubusercontent.com/80676553/111132121-f95ed200-859e-11eb-9ce7-f53011d948b4.png">

**Comment by marefr** [comment]
Code is here and seems to handle alert and close separately currently: https://github.com/grafana/grafana/blob/e165eda283cd9bdd0b98902782759619d52d14b2/pkg/services/alerting/notifiers/opsgenie.go#L96-L100

**Comment by marefr** [comment]
Seems like the implementation are posting to a certain endpoint for closing that alert (earlier triggered). Seems weird to me any other information should be needed?
https://github.com/grafana/grafana/blob/e165eda283cd9bdd0b98902782759619d52d14b2/pkg/services/alerting/notifiers/opsgenie.go#L181

**Comment by RanjithPremanand** [comment]
Whenever there is an alert triggered from grafana to opegenie, it creates an alert in opsgenie based on the create alert action setup as expected
but when the alert goes back to normal state, ideally the opsgenie alert should also get closed based on the close alert action setup, but it is not doing so.
the reason we found is that from grafana it always emits alias id as the endpoint for closing the alert @marefr @wbrowne

**Comment by RanjithPremanand** [comment]
https://github.com/grafana/grafana/issues/31295 and this is the open issue if we just keep the alias id as endpoint @wbrowne @marefr

**Comment by RanjithPremanand** [comment]
so we changed that to a url instead of a alias so that it will be unique
now even if we change that in opsgenie, grafana still sends alias id as a search match for closing the alert due to some existing code
now my requirement can that be changed in a way that the other fields like url or description can be compared to match instead just searching for alias @marefr @wbrowne

**Comment by RanjithPremanand** [comment]
cmd := &models.SendWebhookSync{
		Url:        fmt.Sprintf("%s/alertId-%d/close?identifierType=alias", on.APIUrl, evalContext.Rule.ID),
		Body:       string(body),
		HttpMethod: "POST",
		HttpHeader: map[string]string{
			"Content-Type":  "application/json",
			"Authorization": fmt.Sprintf("GenieKey %s", on.APIKey),

i doubt the above code from your end is causing this issue

can that be changed? @marefr @wbrowne

**Comment by RanjithPremanand** [comment]
> Seems like the implementation are posting to a certain endpoint for closing that alert (earlier triggered). Seems weird to me any other information should be needed?
> https://github.com/grafana/grafana/blob/e165eda283cd9bdd0b98902782759619d52d14b2/pkg/services/alerting/notifiers/opsgenie.go#L181

What this piece of code actually does? @marefr

**Comment by marefr** [comment]
I'm sorry but your latest comments made me even more confused. Let's back up a bit.

What does #31295 has to do with this? Are you asking for the alias defined, see below, to be changed to something else?
https://github.com/grafana/grafana/blob/4aff0cee0dcee7d9f59e6a9a6b53009d6b48aea8/pkg/services/alerting/notifiers/opsgenie.go#L160
https://github.com/grafana/grafana/blob/4aff0cee0dcee7d9f59e6a9a6b53009d6b48aea8/pkg/services/alerting/notifiers/opsgenie.go#L221

I thought this issue was regarding your alerts didn't close in opsgenie when resolved in Grafana. Is that still true? 

What I was trying to explain is that the following JSON payload is provided when creating alerts in opsgenie:
https://github.com/grafana/grafana/blob/4aff0cee0dcee7d9f59e6a9a6b53009d6b48aea8/pkg/services/alerting/notifiers/opsgenie.go#L157-L161

When closing an alert the following JSON payload is provided to opsgenie:
https://github.com/grafana/grafana/blob/4aff0cee0dcee7d9f59e6a9a6b53009d6b48aea8/pkg/services/alerting/notifiers/opsgenie.go#L216-L218

But related I provided the code for the URL being used when closing an alert in opsgenie and that's include the alias in the URL path followed by close and also has the querystring identifierType=alias. Given all this information I would image that should close the alert in opsgenie given the provided alias. But perhaps something has changed in their API?
https://github.com/grafana/grafana/blob/4aff0cee0dcee7d9f59e6a9a6b53009d6b48aea8/pkg/services/alerting/notifiers/opsgenie.go#L221

Are you sure that you've enabled the `Auto close incidents` option for the notifier?

Anyhow, this is not something maintainers are planning to work on and would need help from contributors.

**Comment by lionbee** [comment]
We use Grafana with opsgenie and we don't have this issue. As @marefr pointed out above, in the Notification channel settings this behaviour needs to be configured.

![image](https://user-images.githubusercontent.com/3923816/114508722-0b07c800-9c78-11eb-9ebe-808455054139.png)

**Comment by RanjithPremanand** [comment]
@lionbee Yes, this setting is enabled in our setup too. Does your setup have multiple instances of grafana running which is connected to opsgenie?

**Comment by lionbee** [comment]
@RanjithPremanand no, we have a single instance of Grafana for this particular use case.

**Comment by RanjithPremanand** [comment]
> @RanjithPremanand no, we have a single instance of Grafana for this particular use case.

@lionbee It works well for single grafana instance. But in our setup its multiple instances of grafana where we face this issue

**Comment by natellium** [comment]
Hey there! We want to let you all know that we are moving to Github Issues for feature requests. You can read more about the move [here](https://github.com/grafana/grafana/discussions/73424).

I will be closing this discussion, but if you’re still interested in this request being implemented please don’t hesitate to open a [feature request](https://github.com/grafana/grafana/issues/new?assignees=&labels=&projects=&template=1-feature_requests.md). Sorry for the inconvenience, and thanks for your contributions!

cc @grafana/alerting-squad
